File: Demo/src/main/java/com/braintreepayments/demo/MainFragment.java
package com.braintreepayments.demo;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;

import com.braintreepayments.api.PaymentMethodNonce;

public class MainFragment extends Fragment {

    static final String EXTRA_PAYMENT_RESULT = "payment_result";
    static final String EXTRA_DEVICE_DATA = "device_data";
    static final String EXTRA_COLLECT_DEVICE_DATA = "collect_device_data";

    private static final String KEY_NONCE = "nonce";

    private PaymentMethodNonce nonce;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_main, container, false);

        Button googlePayButton = view.findViewById(R.id.google_pay);
        Button cardsButton = view.findViewById(R.id.card);
        Button payPalButton = view.findViewById(R.id.paypal);
        Button venmoButton = view.findViewById(R.id.venmo);
        Button visaCheckoutButton = view.findViewById(R.id.visa_checkout);
        Button localPaymentsButton = view.findViewById(R.id.local_payment);
        Button preferredPaymentMethodsButton = view.findViewById(R.id.preferred_payment_methods);
        Button samsungButton = view.findViewById(R.id.samsung_pay);
        Button sepaDirectDebitButton = view.findViewById(R.id.sepa_debit);
        Button payPalNativeCheckoutButton = view.findViewById(R.id.paypal_native_checkout);

        cardsButton.setOnClickListener(this::launchCards);
        payPalButton.setOnClickListener(this::launchPayPal);
        localPaymentsButton.setOnClickListener(this::launchLocalPayment);
        googlePayButton.setOnClickListener(this::launchGooglePay);
        visaCheckoutButton.setOnClickListener(this::launchVisaCheckout);
        venmoButton.setOnClickListener(this::launchVenmo);
        preferredPaymentMethodsButton.setOnClickListener(this::launchPreferredPaymentMethods);
        samsungButton.setOnClickListener(this::launchSamsungPay);
        payPalNativeCheckoutButton.setOnClickListener(this::launchPayPalNativeCheckout);
        sepaDirectDebitButton.setOnClickListener(this::launchSEPADirectDebit);

        return view;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(KEY_NONCE)) {
                nonce = savedInstanceState.getParcelable(KEY_NONCE);
            }
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (nonce != null) {
            outState.putParcelable(KEY_NONCE, nonce);
        }
    }

    public void launchGooglePay(View v) {
        NavDirections action =
                MainFragmentDirections.actionMainFragmentToGooglePayFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchCards(View v) {
        Bundle args = new Bundle();
        args.putBoolean(EXTRA_COLLECT_DEVICE_DATA, Settings.shouldCollectDeviceData(getActivity()));

        MainFragmentDirections.ActionMainFragmentToCardFragment action =
                MainFragmentDirections.actionMainFragmentToCardFragment();
        action.setShouldCollectDeviceData(Settings.shouldCollectDeviceData(getActivity()));

        Navigation.findNavController(v).navigate(action);
    }

    public void launchPayPal(View v) {
        MainFragmentDirections.ActionMainFragmentToPayPalFragment action =
                MainFragmentDirections.actionMainFragmentToPayPalFragment();
        action.setShouldCollectDeviceData(Settings.shouldCollectDeviceData(getActivity()));

        Navigation.findNavController(v).navigate(action);
    }

    public void launchVenmo(View v) {
        NavDirections action = MainFragmentDirections.actionMainFragmentToVenmoFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchVisaCheckout(View v) {
        NavDirections action = MainFragmentDirections.actionMainFragmentToVisaCheckoutFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchLocalPayment(View v) {
        NavDirections action =
                MainFragmentDirections.actionMainFragmentToLocalPaymentFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchPreferredPaymentMethods(View v) {
        NavDirections action =
                MainFragmentDirections.actionMainFragmentToPreferredPaymentMethodsFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchSamsungPay(View v) {
        NavDirections action =
                MainFragmentDirections.actionMainFragmentToSamsungPayFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchPayPalNativeCheckout(View v) {
        NavDirections action = MainFragmentDirections.actionMainFragmentToPayPalNativeCheckoutFragment();
        Navigation.findNavController(v).navigate(action);
    }

    public void launchSEPADirectDebit(View v) {
        NavDirections action = MainFragmentDirections.actionMainFragmentToSepaDirectDebitFragment();
        Navigation.findNavController(v).navigate(action);
    }
}

File: Demo/src/main/java/com/braintreepayments/demo/MainFragment.kt
package com.braintreepayments.demo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.navigation.fragment.findNavController
import com.braintreepayments.api.core.PaymentMethodNonce
import com.braintreepayments.api.sharedutils.IntentExtensions.parcelable

@Suppress("TooManyFunctions")
class MainFragment : BaseFragment() {

    private var nonce: PaymentMethodNonce? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return ComposeView(requireContext()).apply {
            setContent {
                Column(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {

                    Card(
                        modifier = Modifier.padding(12.dp),
                    ) {
                        Text(
                            modifier = Modifier
                                .align(Alignment.CenterHorizontally)
                                .padding(top = 8.dp),
                            text = "PayPal",
                            style = MaterialTheme.typography.headlineSmall
                        )

                        PaymentModuleButton(R.string.paypal_button) { launchPayPal() }
                        PaymentModuleButton(R.string.shopper_insights_button) { launchShopperInsights() }
                        PaymentModuleButton(R.string.paypal_messaging_button) { launchPayPalMessaging() }
                        Spacer(modifier = Modifier.height(4.dp))
                    }

                    Card(
                        modifier = Modifier.padding(12.dp),
                    ) {
                        Text(
                            modifier = Modifier
                                .align(Alignment.CenterHorizontally)
                                .padding(top = 8.dp),
                            text = "Other",
                            style = MaterialTheme.typography.headlineSmall
                        )

                        PaymentModuleButton(R.string.venmo) { launchVenmo() }
                        PaymentModuleButton(R.string.cards) { launchCards() }
                        PaymentModuleButton(R.string.google_pay) { launchGooglePay() }
                        PaymentModuleButton(R.string.visa_checkout_button) { launchVisaCheckout() }
                        PaymentModuleButton(R.string.local_payment_button) { launchLocalPayment() }
                        PaymentModuleButton(R.string.sepa_direct_debit_button) { launchSEPADirectDebit() }
                        Spacer(modifier = Modifier.height(4.dp))
                    }
                }
            }
        }
    }

    @Composable
    private fun PaymentModuleButton(
        stringResource: Int,
        onClick: () -> Unit,
    ) {
        Button(
            modifier = Modifier
                .padding(horizontal = 12.dp, vertical = 4.dp)
                .fillMaxWidth(),
            onClick = onClick
        ) {
            Text(text = stringResource(stringResource))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState != null && savedInstanceState.containsKey(KEY_NONCE)) {
            nonce = savedInstanceState.parcelable(KEY_NONCE)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (nonce != null) {
            outState.putParcelable(KEY_NONCE, nonce)
        }
    }

    private fun fetchAuthorizationAndHandleError(onSuccess: (String) -> Unit) {
        fetchAuthorization { authResult ->
            when (authResult) {
                is BraintreeAuthorizationResult.Success -> onSuccess(authResult.authString)
                is BraintreeAuthorizationResult.Error -> showDialog(authResult.error.toString())
            }
        }
    }

    private fun launchGooglePay() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToGooglePayFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchCards() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToCardFragment()
            action.setShouldCollectDeviceData(Settings.shouldCollectDeviceData(requireContext()))
            action.setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchPayPal() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToPayPalFragment()
            action.setShouldCollectDeviceData(Settings.shouldCollectDeviceData(requireContext()))
            action.setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchVenmo() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToVenmoFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchVisaCheckout() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToVisaCheckoutFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchLocalPayment() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToLocalPaymentFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchSEPADirectDebit() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToSepaDirectDebitFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchPayPalMessaging() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToPayPalMessagingFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    private fun launchShopperInsights() {
        fetchAuthorizationAndHandleError { authString ->
            val action = MainFragmentDirections.actionMainFragmentToShoppingInsightsFragment()
                .setAuthString(authString)
            findNavController().navigate(action)
        }
    }

    companion object {
        const val EXTRA_COLLECT_DEVICE_DATA: String = "collect_device_data"

        private const val KEY_NONCE = "nonce"
    }
}
